src/libsync/creds/oauth.cpp

```diff
@@ -182,25 +176,40 @@ void OAuth::startAuthentification()
-        QObject::connect(job, &SimpleNetworkJob::finishedSignal, this, [this](QNetworkReply *reply) {
-            auto jsonData = reply->readAll();
+        connect(job, &SimpleNetworkJob::finishedSignal, this, [this, refreshToken](QNetworkReply *reply) {
+            const auto jsonData = reply->readAll();
+            QString accessToken;
+            QString newRefreshToken = refreshToken;
             QJsonParseError jsonParseError;
+            // https://developer.okta.com/docs/reference/api/oidc/#response-properties-2
             const QJsonObject json = QJsonDocument::fromJson(jsonData, &jsonParseError).object();
-            const QString accessToken = json["access_token"].toString();
-            if (jsonParseError.error != QJsonParseError::NoError || json.isEmpty()) {
-                // Invalid or empty JSON: Network error maybe?
-                qCWarning(lcOauth) << "Error while refreshing the token" << reply->errorString() << jsonData << jsonParseError.errorString();
-            } else if (accessToken.isEmpty()) {
-                // If the json was valid, but the reply did not contain an access token, the token
-                // is considered expired. (Usually the HTTP reply code is 400)
-                qCDebug(lcOauth) << "Expired refresh token. Logging out";
-                Q_EMIT refreshFinished(QString(), QString());
+            const QString error = json.value(QLatin1String("error")).toString();
+            if (!error.isEmpty()) {
+                if (error == QLatin1String("invalid_grant") ||
+                    error == QLatin1String("invalid_request")) {
+                    newRefreshToken.clear();
+                } else {
+                    qCWarning(lcOauth) << tr("Error while refreshing the token: %1 : %2").arg(error, json.value(QLatin1String("error_description")).toString());
+                }
+            } else if (reply->error() != QNetworkReply::NoError) {
+                qCWarning(lcOauth) << tr("Error while refreshing the token: %1 : %2").arg(reply->errorString(), QString::fromUtf8(jsonData));
             } else {
-                Q_EMIT refreshFinished(accessToken, json["refresh_token"].toString());
+                if (jsonParseError.error != QJsonParseError::NoError || json.isEmpty()) {
+                    // Invalid or empty JSON: Network error maybe?
+                    qCWarning(lcOauth) << tr("Error while refreshing the token: %1 : %2").arg(jsonParseError.errorString(), QString::fromUtf8(jsonData));
+                } else {
+                    QString error;
+                    accessToken = getRequiredField(json, QStringLiteral("access_token"), &error).toString();
+                    newRefreshToken = getRequiredField(json, QStringLiteral("refresh_token"), &error).toString();
+                    if (!error.isEmpty()) {
+                        qCWarning(lcOauth) << tr("The reply from the server did not contain all expected fields\n:%1").arg(error);
```

Comment:
We need to be carefull not to get in an infinite loop that would flood the server with token refresh queries.

We probably should if the server is returning another error for a reason or another, that would happen.